lago/subnet_lease.py

```diff
@@ -17,222 +17,505 @@
+                the bounds).
+        """
+        return '{} - {}'.format(self._min_subnet, self._max_subnet)
+
+    @property
+    def path(self):
+        return self._path
+
+
+class Lease(object):
     """
-    Lease a free network for the given uuid path
-
-    Args:
-        uuid_path (str): Path to the uuid file of a :class:`lago.Prefix`
+    Lease object is an abstraction of a lease file.
 
-    Returns:
-        str: the dotted ip of the gateway for the leased net
-
-    .. todo:: _aquire might return None, this will throw a TypeError
+    Attributes:
+        _store_path (str): Path to the lease's store.
+        _subnet (str): The subnet that this lease represents
+        _path (str): The path to the lease file
     """
-    return '192.168.%d.1' % _acquire(uuid_path)
-
 
-@_validate_lease_dir_present
-@_locked
-def _release(index):
-    """
-    Free the lease of the given subnet index
-
-    Args:
-        index (int): Third element of a dotted ip representation of the subnet,
-            for example, for 1.2.3.4 it would be 3
-
-    Returns:
-        None
-    """
-
-    lease_dir = config['lease_dir']
-    lease_file = os.path.join(lease_dir, '%d.lease' % index)
-    os.unlink(lease_file)
-
-
-def release(subnet):
-    """
-    Free the lease of the given subnet
-
-    Args:
-        subnet (str): dotted ip or network to free the lease of
-
-    Returns:
-        None
-    """
-    _release(int(subnet.split('.')[2]))
+    def __init__(self, store_path, subnet):
+        self._store_path = store_path
+        self._subnet = subnet
+        self._path = None
+        self._realise_lease_path()
+
+    @staticmethod
+    def from_lease_filename(store_path, store_template, lease_filename):
+        third_octet = lease_filename.split('.')[0]
+        return Lease(store_path, store_template.format(third_octet))
+
+    def _realise_lease_path(self):
+        ip = self.subnet.split('/')[0]
+        idx = ip.split('.')[2]
+        self._path = os.path.join(self._store_path, '{}.lease'.format(idx))
+
+    def to_ip_network(self):
+        return IPNetwork(self.subnet)
+
+    @property
+    def valid(self):
+        if self.exist:
+            return self.has_env
+        else:
+            return False
+
+    @property
+    def metadata(self):
+        with open(self.path) as f:
+            uuid_path, uuid = json.load(f)
+
+        return uuid_path, uuid
+
+    @property
+    def has_env(self):
+        return self._has_env()
+
+    def _has_env(self, uuid_path=None, uuid=None):
+        if not (uuid_path and uuid):
+            uuid_path, uuid = self.metadata
+
+        if not os.path.isfile(uuid_path):
+            return False
+
+        with open(uuid_path, mode='rt') as f:
+            if f.read() == uuid:
+                return uuid_path
+            else:
+                return False
+
+    @property
+    def exist(self):
+        return os.path.isfile(self.path)
```

Comment:
This will throw a nasty exception if the user does not have read access to that path.
In the longer run we should discuss how can we re-factor it to support multi-users..

For now I think adding a try/except here and assuming that if you don't have read access it just exists would be fine, no?